File: src/baseDatos/Serializador.java
package baseDatos;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.ObjectOutputStream;
import java.io.PrintWriter;

import gestorAplicacion.gestionZoologico.Administracion;

public class Serializador {
	// Este atributo es para definir la ruta al directorio temp que contiene las clases.
	private static File rutaTemp = new File("src\\baseDatos\\temp");
	
	// El método serializar(...) es el encargado de serializar las listas de todos los objetos que están en la clase Administracion
	public static void serializar() {
		FileOutputStream fos;
		ObjectOutputStream oos;
		File [] archivos = rutaTemp.listFiles();
		PrintWriter pw;
		
		/* El siguiente ciclo for borra el contenido de los archivos al momento de guardar los objetos, esto para evitar que haya
		 * reduncancia en los archivos y futuras complicaciones para buscar. */
		for(File archivo : archivos) {
			try {
				/* Al crear el siguiente objeto tipo PrintWriter y pasarle como parámetro la ruta de cada archivo, borrará lo
				 * que haya en cada archivo automáticamente. */
				pw = new PrintWriter(archivo);
			} catch (FileNotFoundException excepcion) {
				excepcion.printStackTrace();
			}
		}
		
		for(File archivo : archivos) {
			if (archivo.getAbsolutePath().contains("administracion")) {
				try {
					fos = new FileOutputStream(archivo);
					oos = new ObjectOutputStream(fos);
					oos.writeObject(Administracion.getCaja());
				} catch(FileNotFoundException excepcion) {
					excepcion.printStackTrace();
				} catch(IOException excepcion) {
					excepcion.printStackTrace();
				}
			} else if (archivo.getAbsolutePath().contains("animales")) {
				try {
					fos = new FileOutputStream(archivo);
					oos = new ObjectOutputStream(fos);
					oos.writeObject(Administracion.getAnimales());
				} catch(FileNotFoundException excepcion) {
					excepcion.printStackTrace();
				} catch(IOException excepcion) {
					excepcion.printStackTrace();
				}
			} else if(archivo.getAbsolutePath().contains("cuidadores")) {
				try {
					fos = new FileOutputStream(archivo);
					oos = new ObjectOutputStream(fos);
					oos.writeObject(Administracion.getCuidadores());
				} catch(FileNotFoundException excepcion) {
					excepcion.printStackTrace();
				} catch(IOException excepcion) {
					excepcion.printStackTrace();
				}
			} else if(archivo.getAbsolutePath().contains("habitats")) {
				try {
					fos = new FileOutputStream(archivo);
					oos = new ObjectOutputStream(fos);
					oos.writeObject(Administracion.getHabitats());
				} catch(FileNotFoundException excepcion) {
					excepcion.printStackTrace();
				} catch(IOException excepcion) {
					excepcion.printStackTrace();
				}
			} else if(archivo.getAbsolutePath().contains("veterinarios")) {
				try {
					fos = new FileOutputStream(archivo);
					oos = new ObjectOutputStream(fos);
					oos.writeObject(Administracion.getVeterinarios());
				} catch(FileNotFoundException excepcion) {
					excepcion.printStackTrace();
				} catch(IOException excepcion) {
					excepcion.printStackTrace();
				}
			} else if(archivo.getAbsolutePath().contains("visitantes")) {
				try {
					fos = new FileOutputStream(archivo);
					oos = new ObjectOutputStream(fos);
					oos.writeObject(Administracion.getVisitantes());
				} catch(FileNotFoundException excepcion) {
					excepcion.printStackTrace();
				} catch(IOException excepcion) {
					excepcion.printStackTrace();
				}
			}
		}
	}
}

File: Python/src/baseDatos/Serializador.py
"""Serializador creado por: JOSÉ DAVID CARDONA SOTO
Este módulo fue hecho con la finalidad de crear la función para serilizar los objetos creados, para
así luego poder ser deserializados y usados en el main de nuestra aplicación. Cada tipo de objeto
es serializado por aparte, en su función correspondiente, y luego invocado en la función serializar"""

from gestorAplicacion.administracion import Administracion
import os
import pathlib

import pickle

#Se llaman a todas las funciones de serilización
def serializar():
    serializarAdmin()
    serializarAnimales()
    serializarCuidadores()
    serializarVeterinarios()
    serializarHabitats()
    serializarVisitantes()
    
def serializarAdmin():
    #Creación y apertura del archivo donde será guardado el flujo de bytes que representen el objeto. En este caso se trata de administracion
    fichero_admin=open(os.path.join(pathlib.Path(__file__).parent.absolute(), "temp\\administracion"),"wb") 
    #Indicamos el dato que será serializado. En este caso el valor de la caja de Administración
    pickle.dump(Administracion.getCaja(),fichero_admin)
    #Se cierra el archivo creado
    fichero_admin.close()

def serializarAnimales():
    #Creación y apertura del archivo donde será guardado el flujo de bytes que representen el objeto. En este caso se trata de animales
    fichero_animales=open(os.path.join(pathlib.Path(__file__).parent.absolute(), "temp\\animales"),"wb")
    #Indicamos el dato que será serializado. En este caso la lista de animales del zoológico
    pickle.dump(Administracion.getAnimales(),fichero_animales)
    #Se cierra el archivo creado
    fichero_animales.close()

def serializarCuidadores():
    #Creación y apertura del archivo donde será guardado el flujo de bytes que representen el objeto. En este caso se trata de cuidadores
    fichero_cuidadores=open(os.path.join(pathlib.Path(__file__).parent.absolute(), "temp\\cuidadores"),"wb")
    #Indicamos el dato que será serializado. En este caso la lista de cuidadores del zoológico
    pickle.dump(Administracion.getCuidadores(),fichero_cuidadores)
    #Se cierra el archivo creado
    fichero_cuidadores.close()

def serializarVeterinarios():
    #Creación y apertura del archivo donde será guardado el flujo de bytes que representen el objeto. En este caso se trata de veterinarios
    fichero_veterinarios=open(os.path.join(pathlib.Path(__file__).parent.absolute(), "temp\\veterinarios"),"wb")
    #Indicamos el dato que será serializado. En este caso la lista de veterinarios del zoológico
    pickle.dump(Administracion.getVeterinarios(),fichero_veterinarios)
    #Se cierra el archivo creado
    fichero_veterinarios.close()

def serializarHabitats():
    #Creación y apertura del archivo donde será guardado el flujo de bytes que representen el objeto. En este caso se trata de habitats
    fichero_habitats=open(os.path.join(pathlib.Path(__file__).parent.absolute(), "temp\\habitats"),"wb")
    #Indicamos el dato que será serializado. En este caso la lista de habitats del zoológico
    pickle.dump(Administracion.getHabitats(),fichero_habitats)
    #Se cierra el archivo creado
    fichero_habitats.close()

def serializarVisitantes():
    #Creación y apertura del archivo donde será guardado el flujo de bytes que representen el objeto. En este caso se trata de visitantes
    fichero_visitantes=open(os.path.join(pathlib.Path(__file__).parent.absolute(), "temp\\visitantes"),"wb")
    #Indicamos el dato que será serializado. En este caso la lista de visitants del zoológico
    pickle.dump(Administracion.getVisitantes(),fichero_visitantes)
    #Se cierra el archivo creado
    fichero_visitantes.close()

